Sources/SwiftSyntax/SyntaxTransform.swift.gyb

```diff
@@ -0,0 +1,122 @@
+%{
+  from gyb_syntax_support import *
+  # -*- mode: Swift -*-
+  # Ignore the following admonition it applies to the resulting .swift file only
+}%
+//// Automatically Generated From SyntaxTransform.swift.gyb.
+//// Do Not Edit Directly!
+//===------ SyntaxTransform.swift - Syntax Transform Visitor Protocol -----===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+public protocol SyntaxTransformVisitor {
+  associatedtype ResultType = Void
+  
+  func visitAny(_ node: Syntax) -> ResultType
+  
+  func visit(_ token: TokenSyntax) -> ResultType
+  func visit(_ node: UnknownSyntax) -> ResultType
+  
+% for node in SYNTAX_NODES:
+%   if is_visitable(node):
+  /// Visiting `${node.name}` specifically.
+  ///   - Parameter node: the node we are visiting.
+  ///   - Returns: the sum of whatever the child visitors return.
+  func visit(_ node: ${node.name}) -> ResultType
+%   end
+% end
+}
+
+extension SyntaxTransformVisitor {
+  public func visit(_ token: TokenSyntax) -> ResultType { visitAny(Syntax(token)) }
+  public func visit(_ node: UnknownSyntax) -> ResultType { visitAny(Syntax(node)) }
+
+% for node in SYNTAX_NODES:
+%   if is_visitable(node):
+  /// Visiting `${node.name}` specifically.
+  ///   - Parameter node: the node we are visiting.
+  ///   - Returns: nil by default.
+  public func visit(_ node: ${node.name}) -> ResultType {
+      visitAny(Syntax(node))
+  }
+%   end
+% end
+
+  public func visit(_ data: Syntax) -> ResultType {
+    switch data.raw.kind {
+    case .token:
+      let node = TokenSyntax(data)!
+      return visit(node)
+    case .unknown:
+      let node = UnknownSyntax(data)!
+      return visit(node)
+    // The implementation of every generated case goes into its own function. This
+    // circumvents an issue where the compiler allocates stack space for every
+    // case statement next to each other in debug builds, causing it to allocate
+    // ~50KB per call to this function. rdar://55929175
+  % for node in NON_BASE_SYNTAX_NODES:
+    case .${node.swift_syntax_kind}:
+      let node = ${node.name}(data)!
+      return visit(node)
+  % end
+    }
+  }
+  
+  public func visit(_ data: ExprSyntax) -> ResultType {
+    switch data.raw.kind {
+  % for node in NON_BASE_SYNTAX_NODES:
+  %   if node.base_kind == "Expr":
+    case .${node.swift_syntax_kind}:
+      let node = data.as(${node.name}.self)!
+      return visit(node)
+  %   end
+  % end
+    default:
+      fatalError("Not expression?")
+    }
+  }
```

Comment:
Why do we need this method? I don’t think we should ever end up calling `visit` with a base node type. Same for the methods below.